### PR TITLE
Disallow Solr in WordPress admin unless coming from admin ajax

### PR DESCRIPTION
**Pull request by ataylorme**
Addresses #215

### CODE REVIEW

**Comment by ataylorme** [comment]
@joshkoenig or @danielbachhuber mind giving this a review? See #215 for background

File: includes/class-solrpower.php [hunk 1]
```diff
@@ -251,8 +255,8 @@ function ajax_search() {
 		} );
 
 		// Allow an AJAX search.
-		add_filter( 'solr_allow_ajax', '__return_true' );
-		add_filter( 'solr_allow_admin', '__return_true' );
+		add_filter( 'solr_allow_ajax', array( $this, 'return_true' ) );
+		add_filter( 'solr_allow_admin', array( $this, 'return_true' ) );
```
Comment (danielbachhuber) [inline includes/class-solrpower.php]: Why create a dedicated method for this?

File: includes/class-solrpower-wp-query.php [hunk 2]
```diff
@@ -81,8 +81,8 @@ function setup() {
 		 *
 		 * @param bool $solr_allow_admin True to query in WordPress Dashboard or false [default false].
 		 */
-		if ( is_admin() && false === apply_filters( 'solr_allow_admin', false ) ) {
-			// return;
+		if ( is_admin() && false === apply_filters( 'solr_allow_admin', false ) && (!defined( 'DOING_AJAX' ) || !DOING_AJAX ) ) {
```
Comment (danielbachhuber) [inline includes/class-solrpower-wp-query.php]: This:

```
(!defined( 'DOING_AJAX' ) || !DOING_AJAX )
```

Would be better expressed as:

```
! ( defined( 'DOING_AJAX' ) && DOING_AJAX )
```

**Comment by ataylorme** [comment]
> My only question is whether we have other use-cases where we want to allow solr power to take over on the backend. If not, wouldn't it be simpler just to disengage under any circumstance where is_admin() is true?

@joshkoenig there's an existing filter for `solr_allow_admin` which defaults to `false`. I see no harm in leaving it for backwards compatibility. It also gives us the option to do Solr in the admin correctly in the future.

File: includes/class-solrpower.php (hunk 1 above)
Comment (ataylorme) [inline includes/class-solrpower.php]: @danielbachhuber it was the end of the day and I brain farted thinking about anonymous functions and how they are a pain to remove once added. `__return_true` is a standard PHP function so you're right we should be using it.

File: includes/class-solrpower-wp-query.php (hunk 2 above)
Comment (ataylorme) [inline includes/class-solrpower-wp-query.php]: Fixed in the latest commit

**Comment by ataylorme** [comment]
@joshkoenig since Daniel is out on vacation can you approve the PR changes? I addressed the items in his code review and would like to merge.

File: includes/class-solrpower.php [hunk 3]
```diff
@@ -250,9 +250,9 @@ function ajax_search() {
 			return $url;
 		} );
 
-		// Allow an AJAX search.
-		add_filter( 'solr_allow_ajax', '__return_true' );
-		add_filter( 'solr_allow_admin', '__return_true' );
+		// Add __return_true to the admin/ajax filters so the widget can access admin ajax.
+		add_filter( 'solr_allow_ajax', array( $this, '__return_true' ) );
+		add_filter( 'solr_allow_admin', array( $this, '__return_true' ) );
```
Comment (danielbachhuber) [inline includes/class-solrpower.php]: Not quite correct either. Should be:

```
add_filter( 'solr_allow_ajax', '__return_true' );
```

This should've failed the tests, so it'd be good to add a test that fails with this invalid callback.

Comment (ataylorme) [inline includes/class-solrpower.php]: d'oh - forgot to remove when I removed the class method.

Comment (ataylorme) [inline includes/class-solrpower.php]: @danielbachhuber can you review one more time?

**Pull request merged**